### PR TITLE
added Makefile

### PR DESCRIPTION
**Pull request by ahalim-intel**
replacing build script build.sh with GNUMake file. This will allow
one script to build, test, run go tool such as lint fmt and build
Docker image.

Change-Id: Ic4bc7db14e5e276902daaeefe13a6a55277d57eb

### CODE REVIEW

File: images/README.md [hunk 1]
```diff
@@ -7,8 +7,7 @@ Typically you'd build this from the root of your SR-IOV network device plugin cl
 ```
 $ docker build -t sriov-device-plugin -f ./images/Dockerfile .
 ```
-
-A `build_docker.sh` script is available for building the SR-IOV device plugin docker image from the `./images` directory.
+You can run `make image` to build the docker image as well.
```
Comment (zshi-redhat) [inline images/README.md]: Let's also remove the images/build_docker.sh script.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+					tr '\n' ',')$$pkg \
+			-covermode=$(COVERAGE_MODE) \
+			-coverprofile="$(COVERAGE_DIR)/coverage/`echo $$pkg | tr "/" "-"`.cover" $$pkg ;\
+	 done
+	$Q $(GOCOVMERGE) $(COVERAGE_DIR)/coverage/*.cover > $(COVERAGE_PROFILE)
+	$Q $(GO) tool cover -html=$(COVERAGE_PROFILE) -o $(COVERAGE_HTML)
+	$Q $(GOCOV) convert $(COVERAGE_PROFILE) | $(GOCOVXML) > $(COVERAGE_XML)
+
+.PHONY: lint
+lint: vendor | $(BASE) $(GOLINT) ; $(info  running golint...) @ ## Run golint
```
Comment (zshi-redhat) [inline Makefile]: 'running golint' complains error:
/bin/sh:line1: $(HOME)/sriov-network-device-plugin/bin/golint: No such file or directory.

File: images/README.md (hunk 1 above)
Comment (ahalim-intel) [reply images/README.md]: +1

File: Makefile (hunk 2 above)
Comment (ahalim-intel) [reply Makefile]: Can you run make clean then make?
golint supposed be installed if not found. Please update me on this. Thanks

Comment (zshi-redhat) [reply Makefile]: Tried with make clean;make, still failed.
Seems it's looking for 'golint' under root dir of sriov-network-device-plugin repo : /root/sriov-network-device-plugin/bin/golint, but there is no bin/ dir and golint doesn't exist.

[root@test sriov-network-device-plugin]# make
running gofmt...
setting GOPATH...
building golint...
running golint...
/bin/sh:line1: /root/sriov-network-device-plugin/bin/golint: No such file or directory
github.com/intel/sriov-network-device-plugin/vendor/golang.org/x/net/context
github.com/intel/sriov-network-device-plugin/vendor/github.com/golang/protobuf/proto
github.com/intel/sriov-network-device-plugin/vendor/github.com/golang/glog
github.com/intel/sriov-network-device-plugin/vendor/golang.org/x/net/http2/hpack
github.com/intel/sriov-network-device-plugin/vendor/golang.org/x/text/transform
github.com/intel/sriov-network-device-plugin/vendor/golang.org/x/text/unicode/bidi
github.com/intel/sriov-network-device-plugin/vendor/golang.org/x/text/unicode/norm
github.com/intel/sriov-network-device-plugin/vendor/golang.org/x/net/internal/timeseries
github.com/intel/sriov-network-device-plugin/vendor/golang.org/x/net/trace
github.com/intel/sriov-network-device-plugin/vendor/golang.org/x/text/secure/bidirule
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/grpclog
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/connectivity
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/credentials
github.com/intel/sriov-network-device-plugin/vendor/golang.org/x/net/idna
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/resolver
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/codes
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/balancer
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/internal
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/keepalive
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/metadata
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/naming
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/peer
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/stats
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/tap
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/grpclb/grpc_lb_v1/messages
github.com/intel/sriov-network-device-plugin/vendor/github.com/golang/protobuf/ptypes/any
github.com/intel/sriov-network-device-plugin/vendor/golang.org/x/net/lex/httplex
github.com/intel/sriov-network-device-plugin/vendor/github.com/golang/protobuf/ptypes/duration
github.com/intel/sriov-network-device-plugin/vendor/github.com/golang/protobuf/ptypes/timestamp
github.com/intel/sriov-network-device-plugin/vendor/golang.org/x/net/http2
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/genproto/googleapis/rpc/status
github.com/intel/sriov-network-device-plugin/vendor/github.com/gogo/protobuf/proto
github.com/intel/sriov-network-device-plugin/vendor/github.com/golang/protobuf/ptypes
github.com/intel/sriov-network-device-plugin/vendor/github.com/gogo/protobuf/sortkeys
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/status
github.com/intel/sriov-network-device-plugin/vendor/github.com/gogo/protobuf/protoc-gen-gogo/descriptor
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc/transport
github.com/intel/sriov-network-device-plugin/vendor/github.com/gogo/protobuf/gogoproto
github.com/intel/sriov-network-device-plugin/vendor/google.golang.org/grpc
github.com/intel/sriov-network-device-plugin/vendor/k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1beta1
github.com/intel/sriov-network-device-plugin/cmd/sriovdp
Building sriovdp...
Done!

Comment (ahalim-intel) [reply Makefile]: Yes, it looked for glint binary from usual paths. Then it installed it  in ./bin and proceeded with the build:
```
running gofmt...
setting GOPATH...
building golint...
running golint...
```
Subsequent builds should work as normal. If we continue see the same error then we might need to export the $GOBIN as well.
Also can you also please remove existing ./gotpah directory if its still there? We are moving this into ./.gopath and this directory will be updated on each build. This contains a symlink which may point to an absolute path. And when we are creating docker image we are basically copying everything as is into the Docker contexts.

File: README.md [hunk 3]
```diff
@@ -110,18 +92,18 @@ $ kubectl create -f deployments/crdnetwork.yaml
 $ kubectl create -f deployments/sriov-crd.yaml
 ```
  
- 5. Run build docker script to create SRIOV Network Device Plugin Docker image
+ 5. Build docker image
  ```
-$ cd deployments/
-$ ./build_docker.sh
```
[the remark this answers is not in the document]
Comment (zshi-redhat) [reply README.md]: Cool, I see the images/build_docker.sh file is removed, perhaps we also want to remove deployments/build_docker.sh

File: README.md [hunk 4]
```diff
@@ -60,24 +46,22 @@ The following  NICs were tested with this implementation. However, other SRIOV c
...
+This section explains an exmaple deployment of SRIOV Network device plugin in Kubernetes. Required YAML files can be found in [deployments/](deployments/) directory.
 
 ### Network Object CRDs
-Kubernetes out of the box only allows to have one network interface per pod. In order to add multiple interfaces in a Pod we need to configure Kubernetes with a CNI meta plugin that enables invoking multiple CNI plugins to add additional interfaces.  [Multus](https://github.com/intel/multus-cni) is only meta plugin that supports this mechanism. Multus uses Kubernetes Custom Resource Definition or CRDs to define network objects. For more information see Multus [documentation](https://github.com/intel/multus-cni/blob/master/README.md). 
+Multus uses Custom Resource Definitions(CRDs) for defining additional network attachement. These network attachment CRDs follows the standards defined by K8s Network Plumbing Working Group(NPWG). Please refer to [Multus documentation](https://github.com/intel/multus-cni/blob/master/README.md) for more information.
 
-### Meta Plugin CNI
+### Building Multus
 1. Compile Meta Plugin CNI (Multus dev/k8s-deviceid-model branch):
```
[the remark this answers is not in the document]
Comment (zshi-redhat) [reply README.md]: Now we use Multus master branch for SRIOV DP testing, perhaps remove the content in brackets.

File: README.md [hunk 5]
```diff
@@ -60,24 +46,22 @@ The following  NICs were tested with this implementation. However, other SRIOV c
...
 ````
 $ git clone https://github.com/intel/multus-cni.git
 $ cd multus-cni
-$ git fetch
-$ git checkout dev/k8s-deviceid-model
 $ ./build
 $ cp bin/multus /opt/cni/bin
 ````
 2. Configure Kubernetes network CRD with [Multus](https://github.com/intel/multus-cni/tree/dev/network-plumbing-working-group-crd-change#creating-network-resources-in-kubernetes)
```
[the remark this answers is not in the document]
Comment (ahalim-intel) [reply README.md]: moving creating CRD under Multus and sriov-crd under sriov-cni section. Thanks for the suggestions.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ahalim-intel) [reply README.md]: removed

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+	@docker build -t $(TAG) -f $(DOCKERFILE)  $(CURDIR) 
+endif
+
+# Misc
+
+.PHONY: clean
+clean: ; $(info  Cleaning...)	@ ## Cleanup everything
+	@rm -rf $(GOPATH)
+	@rm -rf $(BUILDDIR)/$(BINARY_NAME)
+	@rm -rf test/tests.* test/coverage.*
```
Comment (zshi-redhat) [inline Makefile]: With $GOBIN dir created for golint build, we'll need to remove the $GOBIN dir when make clean.
for example:  @rm -rf $(GOBIN)
nit: the new update seems add bin/golint dir in the repo which I guess will only be created when make is executed.

Comment (ahalim-intel) [reply Makefile]: the `bin/` directory will contain go-tools executables such as golint,  gocovmerge etc. Should we be removing and adding them on every build? The `bin/` directory is not added to the project repository. So, the make will only create them locally as needed and install required go-tools as well if they are not present.

Comment (zshi-redhat) [reply Makefile]: You're right, the bin/ dir is not added in the repo, and as it's a tool that be used for every build, there is no need to remove it and re-build everytime.

**Comment by zshi-redhat** [comment]
LGTM now, thanks for all the updates!

**Comment by ahalim-intel** [comment]
Thanks for your thorough review @zshi-redhat . Going to merge this now.

**Pull request merged**